service/fx.go

```diff
@@ -91,32 +103,36 @@ func PersistenceMaxQpsFn(
 }
 
 func GrpcServerOptionsProvider(
-	logger log.Logger,
-	rpcFactory common.RPCFactory,
-	retryableInterceptor *interceptor.RetryableInterceptor,
-	telemetryInterceptor *interceptor.TelemetryInterceptor,
-	rateLimitInterceptor *interceptor.RateLimitInterceptor,
-	tracingInterceptor telemetry.ServerTraceInterceptor,
+	params GrpcServerOptionsParams,
 ) []grpc.ServerOption {
 
-	grpcServerOptions, err := rpcFactory.GetInternodeGRPCServerOptions()
+	grpcServerOptions, err := params.RpcFactory.GetInternodeGRPCServerOptions()
 	if err != nil {
-		logger.Fatal("creating gRPC server options failed", tag.Error(err))
+		params.Logger.Fatal("creating gRPC server options failed", tag.Error(err))
 	}
 
 	return append(
 		grpcServerOptions,
-		grpc.ChainUnaryInterceptor(
-			rpc.ServiceErrorInterceptor,
-			grpc.UnaryServerInterceptor(tracingInterceptor),
-			metrics.NewServerMetricsContextInjectorInterceptor(),
-			metrics.NewServerMetricsTrailerPropagatorInterceptor(logger),
-			telemetryInterceptor.UnaryIntercept,
-			rateLimitInterceptor.Intercept,
-			retryableInterceptor.Intercept,
-		),
-		grpc.ChainStreamInterceptor(
-			telemetryInterceptor.StreamIntercept,
-		),
+		grpc.ChainUnaryInterceptor(getUnaryInterceptors(params)...),
+		grpc.ChainStreamInterceptor(params.TelemetryInterceptor.StreamIntercept),
 	)
 }
+
+func getUnaryInterceptors(params GrpcServerOptionsParams) []grpc.UnaryServerInterceptor {
+	interceptors := []grpc.UnaryServerInterceptor{
+		rpc.ServiceErrorInterceptor,
+		grpc.UnaryServerInterceptor(params.TracingInterceptor),
+		metrics.NewServerMetricsContextInjectorInterceptor(),
+		metrics.NewServerMetricsTrailerPropagatorInterceptor(params.Logger),
+		params.TelemetryInterceptor.UnaryIntercept,
+	}
+
+	if len(params.AdditionalInterceptors) > 0 {
```

Comment:
nit: no need to check empty here?

Reply:
Probably not strictly necessary, but it is not a heavily used code path, so I don't mind the extra defensiveness.

Reply:
I'm fine either way but since `foo = append(foo, bar...)` is valid even if `bar` is nil (of slice type), the check here is not really being defensive against anything and may cause confusion (i.e., why check if not needed).